stl/src/atomic.cpp

```diff
@@ -33,4 +39,281 @@ _CRTIMP2_PURE void __cdecl _Unlock_shared_ptr_spin_lock() { // release previousl
+    switch (_Spin_context & _ATOMIC_SPIN_PHASE_MASK) {
+    case _ATOMIC_SPIN_PHASE_INIT_SPIN_COUNT:
+        std::call_once(_Atomic_spin_count_initialized, _Atomic_spin_count_initialize);
+        _Spin_context = _ATOMIC_SPIN_PHASE_INIT_SPIN + _Atomic_spin_count;
+        [[fallthrough]];
+
+    case _ATOMIC_SPIN_PHASE_INIT_SPIN:
+        if ((_Spin_context & _ATOMIC_SPIN_VALUE_MASK) > 0) {
+            _Spin_context -= 1;
+            YieldProcessor();
+            return;
+        }
+        _Spin_context = _ATOMIC_SPIN_PHASE_INIT_SWITCH_THD | (_Spin_context & _ATOMIC_WAIT_PHASE_MASK);
+        [[fallthrough]];
+
+    case _ATOMIC_SPIN_PHASE_INIT_SWITCH_THD:
+        if (_Spin_context < (_ATOMIC_SPIN_PHASE_INIT_SWITCH_THD + 4)) {
+            _Spin_context += 1;
+            ::SwitchToThread();
+            return;
+        }
+        _Spin_context = _ATOMIC_SPIN_PHASE_INIT_SLEEP_ZERO | (_Spin_context & _ATOMIC_WAIT_PHASE_MASK);
+        [[fallthrough]];
+
+    case _ATOMIC_SPIN_PHASE_INIT_SLEEP_ZERO:
+        if (_Spin_context < (_ATOMIC_SPIN_PHASE_INIT_SLEEP_ZERO + 16)) {
+            _Spin_context += 1;
+            ::Sleep(0);
+            return;
+        }
+        _Spin_context = _ATOMIC_SPIN_PHASE_INIT_SLEEP | (_Spin_context & _ATOMIC_WAIT_PHASE_MASK);
+        [[fallthrough]];
+
+    case _ATOMIC_SPIN_PHASE_INIT_SLEEP:
+        ::Sleep(10);
+        return;
+    }
+}
+
+inline bool is_win8_wait_on_address_available() {
+#if _STL_WIN32_WINNT >= _WIN32_WINNT_WIN8
+    return true;
+#else
+    // WaitOnAddress ONLY available on Windows 8+
+    DYNAMICGETCACHEDFUNCTION(PFNWAITONADDRESS, WaitOnAddress, pfWaitOnAddress);
+    return pfWaitOnAddress != nullptr;
+#endif
+}
+
+constexpr size_t TABLE_SIZE_POWER = 8;
+constexpr size_t TABLE_SIZE       = 1 << TABLE_SIZE_POWER;
+constexpr size_t TABLE_MASK       = TABLE_SIZE - 1;
+
+
+#pragma warning(push)
+#pragma warning(disable : 4324) // structure was padded due to alignment specifier
+
+struct alignas(64) _Contention_table_entry {
+    // Arbitraty variable to wait/notify on if target wariable is not proper atomic for that
+    // Size is largest of lock-free to make aliasing problem into hypothetical
+    std::atomic<std::uint64_t> _Counter;
+    // Event to wait on in case of no atomic ops
+    std::atomic<HANDLE> _Semaphore;
+    // Event use count, can delete event if drops to zero
+    // Initialized to one to keep event used when progam runs, will drop to zero on program exit
+    std::atomic<std::size_t> _Semaphore_use_count = 1;
+    // Flag whether semaphore should be released
+    std::atomic<LONG> _Semaphore_own_count = 0;
+    // Once flag for semaphore creation
+    std::once_flag _Semaphore_created;
+    // Once flag for semaphore deletion
+    static std::once_flag _Semaphore_dereference_registered;
+
+    static void _Dereference_all_semaphores();
+
+    HANDLE _Reference_semaphore() {
+        std::call_once(_Semaphore_created, [this] {

```

Comment:
Should probably do CASes for this rather than `call_once`, i.e.

```
static atomic<bool> _Dereference_registered; // add this variable somewhere
// change _Semaphore to start as INVALID_HANDLE_VALUE, that might require making it a uintptr_t rather than a HANDLE to get static initialization

HANDLE _Old = _Semaphore.load(memory_order_acquire);
if (_Old != INVALID_HANDLE_VALUE) { return _Old; }
HANDLE _Created = CreateSemaphore(...); // Try create just once, if low resources, use fall back permanently
if (_Semaphore.compare_exchange_strong(_Old, _Created)) {
    // we did CreateSemaphore first so register the atexit if no thread has done that yet
    if (!_Dereference_registered.load() && !_Dereference_registered.exchange(1)) {
        atexit(_Dereference_all_semaphores);
    }

    return _Created;
} else {
    if (_Created) {
        CloseHandle(_Created); // another thread finished CreateSemaphore first
    }

    return _Old;
}
```

You never need the synchronizing or exception propagating behavior of `call_once` and those aren't cheap.

Reply:
Don't agree with such change for `CreateSemaphore`, although for others looks it fine.
If another thread will call `CreateSemaphore,` it will likely to finish later than first (as first was already halfway by that time). Then it would have to destroy newly created semaphore. Waiting on call_once will be cheaper.

Reply:
Addressed avoiding call_once most of the time. (Still don't go spawning semaphores, instead waiting on call_once)

Reply:
>If another thread will call CreateSemaphore, it will likely to finish later than first (as first was already halfway by that time). Then it would have to destroy newly created semaphore. Waiting on call_once will be cheaper.

True, but this case is extremely unlikely. To be clear, I'm not worried about the cost of call_once in the blocking case as much as the cost of call_once in the not-blocking happy path case.

Reply:
Now we no longer have Semaphore and no longer call_once